File: js/src/formulas/LUDecomposition.ts
export default function LUDecomposition (matrix: number[][]) {
  if (matrix.length !== matrix[0].length) {
    throw new Error()
  }
  const len = matrix.length
  const L = [] as number[][]
  const U = [] as number[][]
  for (let i = 0; i < len; i++) {
    L.push([])
    U.push([])
    for (let j = 0; j < len; j++) {
      L[i].push(0)
      U[i].push(0)
    }
  }

  for (let i = 0; i < len; i++) {
    for (let k = i; k < len; k++) {
      let sum = 0
      for (let j = 0; j < i; j++) {
        sum += L[i][j] * U[j][k]
      }
      U[i][k] = matrix[i][k] - sum
    }
    for (let k = i; k < len; k++) {
      if (i === k) {
        L[i][i] = 1
      } else {
        let sum = 0
        for (let j = 0; j < i; j++) {
          sum += L[k][j] * U[j][i]
        }
        L[k][i] = (matrix[k][i] - sum) / U[i][i]
      }
    }
  }

  return {
    Lower: L,
    Upper: U
  }
}

File: py/src/formulas/LUDecomposition.py
def lu_decomposition(matrix: list[list[float]]):
    if len(matrix) != len(matrix[0]):
        raise Exception()
    length = len(matrix)
    L = [[0 for _ in range(length)] for _ in range(length)]
    U = [[0 for _ in range(length)] for _ in range(length)]
    for i in range(length):
        for k in range(i, length):
            sum_ = 0
            for j in range(i):
                sum_ += L[i][j] * U[j][k]
            U[i][k] = matrix[i][k] - sum_
        for k in range(i, length):
            if i == k:
                L[i][i] = 1
            else:
                sum_ = 0
                for j in range(i):
                    sum_ += L[k][j] * U[j][i]
                L[k][i] = (matrix[k][i] - sum_) / U[i][i]
    return {
        "Lower": L,
        "Upper": U
    }
